### PR TITLE
[CTB-47] Creating and training model

### PR DESCRIPTION
**Pull request by rafalw99**
A StockPredictor (currently using a GRU model) and a StockPredictorManager that allow for prediction of future stock values.

### CODE REVIEW

File: src/server/requirements.txt [hunk 1]
```diff
@@ -3,8 +3,11 @@ flask-cors~=3.0.10
 flask-swagger-ui~=4.11.1
 gunicorn~=20.1.0
 jsonschema~=4.17.3
+numpy~=1.24.2
```
Comment (vis4rd) [inline src/server/requirements.txt:6]: Please create a new `requirements.txt` file in `model/` directory and move packages which are not related to the server.

File: src/server/constants.py [hunk 2]
```diff
@@ -53,3 +53,8 @@ class QUERIES:
                                         WHERE date BETWEEN %s AND %s
                                         GROUP BY period_number
                                         ORDER BY period_number"""
+
+    SELECT_DATA: str = "SELECT date, value FROM exchange_rate_history"
+    SELECT_NEWEST_DATA: str = (
```
Comment (vis4rd) [inline src/server/constants.py:58]: Can these names be changed to something more descriptive, like i.e. `SELECT_ALL_RATE_HISTORY`, `SELECT_ALL_RATE_HISTORY_DESC` etc? These are just suggestions, so feel free to change it or leave it, up2u.

**Review by vis4rd — CHANGES_REQUESTED** [review]

File: src/server/requirements.txt (hunk 1 above)
Comment (rafalw99) [reply src/server/requirements.txt]: Done

File: src/server/constants.py (hunk 2 above)
Comment (rafalw99) [reply src/server/constants.py]: Done

**Review by vis4rd — APPROVED** [review]

**Review by unlvy — APPROVED** [review]

File: src/model/stock_predictor_manager.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    def _get_data(self) -> pd.DataFrame:
+        """Return all data from the stock values table."""
+        data: list = []
+
+        with DatabaseProvider.handler() as handler:
+            handler().execute(QUERIES.SELECT_ALL_RATE_HISTORY)
+            data = handler().fetchall()
+            data = [{"date": date.strftime("%Y-%m-%d"), "avg": avg} for date, avg in data]
+
+        data_frame = pd.DataFrame(data, columns=["date", "value"])
```
Comment (pawel2000pl) [inline src/model/stock_predictor_manager.py:52]: Jako że używamy torch'a, dlaczego nie został użyty jego natywny format (Tensor)?

File: src/model/stock_predictor_manager.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        train_data, val_data = self._get_train_val_data()
+
+        train_loader: DataLoader = self._get_data_loader(train_data)
+        val_loader: DataLoader = self._get_data_loader(val_data)
+
+        no_improvement_count: int = 0
+        best_val_loss: float = sys.float_info.max
+        loss: torch.Tensor = torch.tensor(0)  # pylint: disable=no-member
+
+        for epoch in range(self.config.epochs):
```
Comment (pawel2000pl) [inline src/model/stock_predictor_manager.py:124]: self.config.epochs -> self.config.max_epochs 
może by to zmienić w konfiguracji?

File: src/model/stock_predictor_manager.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+        self.scaler: MinMaxScaler = MinMaxScaler()
+
+        self.criterion: nn.MSELoss = nn.MSELoss()
+        self.optimizer: optim.adam.Adam = optim.Adam(  # pylint: disable=no-member
+            self.stock_predictor.parameters(), lr=self.config.learning_rate
+        )
+
+    def _initialize_model(self) -> None:
+        """Initializing model with parameters from config dataclass."""
+        self.stock_predictor: StockPredictor = StockPredictor(
```
Comment (pawel2000pl) [inline src/model/stock_predictor_manager.py:36]: Brak kompilacji modelu?
(https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html)

File: src/model/stock_predictor_manager.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        while current_date <= date_to_predict:
+            # Scale data and convert it to tensor
+            input_seq_scaled: np.ndarray = self.scaler.fit_transform(data)
+            input_seq_tensor: torch.Tensor = (
+                torch.tensor(input_seq_scaled).float().unsqueeze(0)  # pylint: disable=no-member
+            )
+
+            # Predict output for the next date
+            with torch.no_grad():
+                predicted_output_tensor: torch.Tensor = self.stock_predictor(input_seq_tensor)
```
Comment (pawel2000pl) [inline src/model/stock_predictor_manager.py:187]: Czy nie brakuje self.stock_predictor.eval() ?
- przejście z "trybu" uczenia na "tryb" normalny

File: src/model/stock_predictor_manager.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            ),
+            batch_size=self.config.batch_size,
+        )
+
+        return data_loader
+
+    def train_model(self, verbose: bool = False) -> None:
+        """Train model"""
+
+        train_data: Optional[np.ndarray] = None
```
Comment (pawel2000pl) [inline src/model/stock_predictor_manager.py:113]: Brakuje tu chyba self.stock_predictor.train 
- przejścia modelu w tryb uczenia.

File: src/model/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,4 @@
+numpy~=1.24.2
+pandas~=1.5.3
```
Comment (pawel2000pl) [inline src/model/requirements.txt:2]: Byłbym za usunięciem tej zależności (pandas), skoro dane do uczenia i przewidywania nie są już ładowane z pliku csv.

File: src/model/stock_predictor_manager.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+        self.config = StockPredictorConfig()
+
+        self._initialize_model()
+
+        self.train_size_pct: float = 0.8
+        self.scaler: MinMaxScaler = MinMaxScaler()
+
+        self.criterion: nn.MSELoss = nn.MSELoss()
+        self.optimizer: optim.adam.Adam = optim.Adam(  # pylint: disable=no-member
+            self.stock_predictor.parameters(), lr=self.config.learning_rate
```
[the remark this answers is not in the document]
Comment (rafalw99) [reply src/model/stock_predictor_manager.py]: It turns out it was necessary. I've run several tests, and using the weight_decay parameter gives better results by about 10-15%. Model training is also more stable. It's good to know that such a thing exists.

File: src/model/stock_predictor_manager.py (hunk 4 above)
Comment (rafalw99) [reply src/model/stock_predictor_manager.py]: The variable name has been changed.

File: src/model/stock_predictor_manager.py (hunk 6 above)
Comment (rafalw99) [reply src/model/stock_predictor_manager.py:192]: I forgot to add this earlier, now it has been added.

File: src/model/stock_predictor_manager.py (hunk 7 above)
Comment (rafalw99) [reply src/model/stock_predictor_manager.py:115]: The model's transition to the training mode has been added.

File: src/model/stock_predictor_manager.py (hunk 3 above)
Comment (rafalw99) [reply src/model/stock_predictor_manager.py:54]: Converting directly to the Tensor data type instead of using Pandas as an intermediate type is a good idea. Currently, testing every change related to retrieving data from a database is very difficult. I would suggest pushing the changes to the main branch, integrating the model with the backend, and then removing the use of the Pandas library.

File: src/model/stock_predictor.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
+from typing import Optional
+
+import torch
+from torch import nn
+
+
+class StockPredictor(nn.Module):  # pylint: disable= too-few-public-methods
```
Comment (uNBEx) [inline src/model/stock_predictor.py:7]: `# pylint: disable= too-few-public-methods`

Does our pylint config still enforce that rule? If not, this can be removed.

**Review by uNBEx — CHANGES_REQUESTED** [review]

File: src/model/stock_predictor.py (hunk 10 above)
Comment (rafalw99) [reply src/model/stock_predictor.py]: I wanted to check it, but it seems that checks don't work for this pull request anymore.

Comment (rafalw99) [reply src/model/stock_predictor.py]: "Our pylint doesn't enforce this rule anymore. I removed it.

**Review by pawel2000pl — APPROVED** [review]
